File: LeetCodeAlgos/Python/numberOfIslands.py
# Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.
# An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        res = 0
        for i in range(0, len(grid)):
            for j in range(0, len(grid[0])):
                if grid[i][j] == "1":
                    res+=1
                    self.floodFill(grid, i, j, 0)
        return res

    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        check = image[sr][sc]
        if check == color:
            return image
        image[sr][sc] = color
        if sr > 0 and image[sr-1][sc] == check:
            self.floodFill(image, sr-1, sc, color)
        if sc < len(image[0])-1 and image[sr][sc+1] == check:
            self.floodFill(image, sr, sc+1, color)
        if sr < len(image)-1 and image[sr+1][sc] == check:
            self.floodFill(image, sr+1, sc, color)
        if sc > 0 and image[sr][sc-1] == check:
            self.floodFill(image, sr, sc-1, color)
        return image

File: LeetCodeAlgos/Java/NumberOfIslands.java
// Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

// An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

public class NumberOfIslands {
  public static void main(String[] args) {
    Solution s = new Solution();

    testNumIslands(1, new char[][]{{'1','1','1','1','0'},{'1','1','0','1','0'},{'1','1','0','0','0'},{'0','0','0','0','0'}}, 1, s);
    testNumIslands(2, new char[][]{{'1','1','0','0','0'},{'1','1','0','0','0'},{'0','0','1','0','0'},{'0','0','0','1','1'}}, 3, s);
    testNumIslands(3, new char[][]{{'1'}}, 1, s);

    testNumIslands(4, new char[][]{{'0'}}, 0, s);
  }

  static void testNumIslands(int testNum, char[][] grid, int expected, Solution solution) {
    int result = solution.numIslands(grid);

    System.out.println(String.format(
      "Test %d: %d / %d (%b)",
      testNum,
      result,
      expected,
      result == expected
    ));
  }

}

class Solution {
  public int numIslands(char[][] grid) {
    int count = 0;
      for (int row = 0; row < grid.length; row++) {
        for (int col = 0; col < grid[0].length; col++) {
          if (grid[row][col] == '1') {
            count++;
            cancelIsland(grid, row, col);
          }
        }
      }
      return count;
  }
  public void cancelIsland(char[][] grid, int row, int col) {
    int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    for (int[] direction : directions) {
      int newRow = row + direction[0];
      int newCol = col + direction[1];
      if (newRow >= 0 && newRow < grid.length && newCol >= 0 && newCol < grid[0].length && grid[newRow][newCol] == '1') {
        grid[newRow][newCol] = '0';
        cancelIsland(grid, newRow, newCol);
      }
    }
  }
}
